plugins/inputs/intel_powerstat/intel_powerstat.go

```diff
@@ -0,0 +1,493 @@
+}
+
+func (p *PowerStat) addCoreCPUTemperatureMetric(cpuID string, acc telegraf.Accumulator) {
+	temp := p.msr.GetThrottleTemperature(cpuID) - p.msr.GetTemperature(cpuID)
+	cpu := p.cpuInfo[cpuID]
+	tags := map[string]string{
+		"package_id": cpu.PhysicalID,
+		"core_id":    cpu.CoreID,
+		"cpu_id":     cpu.CPUID,
+	}
+	fields := map[string]interface{}{
+		"name":  "cpu_temperature",
+		"unit":  "celsius_degrees",
+		"value": temp,
+	}
+
+	acc.AddGauge("powerstat_core", fields, tags)
+}
+
+func (p *PowerStat) addProcessorBusyCyclesMetric(cpuID string, acc telegraf.Accumulator) {
+	// Avoid division by 0
+	if p.msr.GetTimestampDelta(cpuID) == 0 {
+		p.Log.Errorf("timestamp counter on offset %s should not equal 0 on cpuID %s",
+			services.TimestampCounterLocation, cpuID)
+		return
+	}
+	busyCyclesValue := services.RoundFloatToNearestTwoDecimalPlaces(percentageMultiplier *
+		float64(p.msr.GetMperfDelta(cpuID)) / float64(p.msr.GetTimestampDelta(cpuID)))
+	cpu := p.cpuInfo[cpuID]
+	tags := map[string]string{
+		"package_id": cpu.PhysicalID,
+		"core_id":    cpu.CoreID,
+		"cpu_id":     cpu.CPUID,
+	}
+	fields := map[string]interface{}{
+		"name":  "cpu_busy_cycles",
+		"unit":  "percentage",
+		"value": busyCyclesValue,
+	}
+
+	acc.AddGauge("powerstat_core", fields, tags)
+}
+
+func (p *PowerStat) addC6StateResidencyMetric(cpuID string, acc telegraf.Accumulator) {
+	// Avoid division by 0
+	if p.msr.GetTimestampDelta(cpuID) == 0 {
+		p.Log.Errorf("timestamp counter on offset %s should not equal 0 on cpuID %s",
+			services.TimestampCounterLocation, cpuID)
+		return
+	}
+	c6Value := services.RoundFloatToNearestTwoDecimalPlaces(percentageMultiplier *
+		float64(p.msr.GetC6Delta(cpuID)) / float64(p.msr.GetTimestampDelta(cpuID)))
+
+	cpu := p.cpuInfo[cpuID]
+	tags := map[string]string{
+		"package_id": cpu.PhysicalID,
+		"core_id":    cpu.CoreID,
+		"cpu_id":     cpu.CPUID,
+	}
+	fields := map[string]interface{}{
+		"name":  "cpu_c6_state_residency",
+		"unit":  "percentage",
+		"value": c6Value,
+	}
+
+	acc.AddGauge("powerstat_core", fields, tags)
+}
+
+func (p *PowerStat) addC1StateResidencyMetric(cpuID string, acc telegraf.Accumulator) {
+	c1Value := 0.0
+	// Avoid division by 0
+	timestampDelta := p.msr.GetTimestampDelta(cpuID)
+	if timestampDelta < 1 {
+		p.Log.Errorf("timestamp delta value %d should not be lower than 1", timestampDelta)
+		return
+	}
+	// Since counter collection is not atomic it may happen that sum of C0, C1, C3, C6 and C7
+	// is bigger value than TSC, in such case C1 residency shall be set to 0.
+	if timestampDelta > (p.msr.GetMperfDelta(cpuID) + p.msr.GetC3Delta(cpuID) +
+		p.msr.GetC6Delta(cpuID) + p.msr.GetC7Delta(cpuID)) {
+		c1 := timestampDelta - p.msr.GetMperfDelta(cpuID) - p.msr.GetC3Delta(cpuID) -
+			p.msr.GetC6Delta(cpuID) - p.msr.GetC7Delta(cpuID)
+
+		c1Value = services.RoundFloatToNearestTwoDecimalPlaces(percentageMultiplier * float64(c1) /
+			float64(timestampDelta))
+	}
```

Comment:
Wouldn't the clipping
```suggestion
	c1 := timestampDelta - p.msr.GetMperfDelta(cpuID) - p.msr.GetC3Delta(cpuID) -
		p.msr.GetC6Delta(cpuID) - p.msr.GetC7Delta(cpuID)
	if c1 < 0.0 {
		c1 = 0.0
	}
	c1Value = services.RoundFloatToNearestTwoDecimalPlaces(percentageMultiplier * float64(c1) /
		float64(timestampDelta))
```
be clearer. As a bonus we avoid calling all the functions twice.

Reply:
Done. And also corrected to avoid overflows.